base/abstractdict.jl

```diff
@@ -465,6 +465,63 @@ function hash(a::AbstractDict, h::UInt)
     hash(hv, h)
 end
 
+"""
+    modify!(f, d::AbstractDict{K, V}, key)
```

Comment:
`modify!` sounds a bit like a weird name to me. `update!` sounds better.

Reply:
FYI, [`update!`](https://github.com/JuliaLang/julia/pull/31367#issuecomment-549182389) and [`set!`](https://github.com/JuliaLang/julia/pull/31367#issuecomment-549100800) were suggested to be simpler shorthands of what I call `modify!` in this PR.  I don't renaming to something else, though.

Reply:
FYI, [`update!`](https://github.com/JuliaLang/julia/pull/31367#issuecomment-549182389) and [`set!`](https://github.com/JuliaLang/julia/pull/31367#issuecomment-549100800) were suggested to be simpler shorthands of what I call `modify!` in this PR.  I don't mind renaming to something else, though.

Reply:
Coming from SQL nomenclature, I think `modify!` is superior to `update!` since this operation can also mean removing an entry... not just updating the value of an existing key.

Reply:
Re-visiting this naming and reading @clarkevans's comment, I actually think `modify!` is still the best name so far.  I think it's better at conveying the point that this API could be used for updating, inserting, or removing an entry.  In particular, `update!` (and with a lesser extent, `set!`) sounds a bit weak at reminding the reader of the code about this.  So, my preference is `modify!` > `set!` >> `update!`.

@nalimilan What do you think?

Reply:
Both "modify" and "update" refer to the dict rather than the value AFAICT, so they both cover the case where you remove the value.

`set!` could be interesting due to the parallel with `get!`. Not sure whether it's a good parallel or not...

Reply:
I prefer `set!` over `modify!` aesthetically, and `modify!` to `update!`.  I think the resistance about `set!` is that `set!(dict, key, nothing)` would remove `key` from `dict`, and hence `set!(dict, key, val) == dict[key]` would hold except when when `val` is `nothing`; but, in this case, `dict[key]` is an error.  Moreover, `set!(dict, key, X) == get(dict, key, X)` even when `X` is `nothing`, correct?  Perhaps I'm not understanding at all.

Reply:
@clarkevans I think @nalimilan mentioned a good point that `set!` sounds like a counter part of `get`.  However, I don't think it's a good parallel.  If you consider `set!` as a counter part of `get(::Container, ::Key) ::Union{Nothing,Some{Value}}`, the signature of `set!` should be `set!(::Container, ::Union{Nothing,Some{Value}}, ::Key)`.  This is how `getindex` and `setindex!` work.

It is conceivable to have this `set!` API since it's very useful when you want to "rollback" a container:

```julia
original = get(dict, key)  # works even if `key` does not exist
dict[key] = tmp
try
    ... do something with dict ...
finally
    # remove the `key` if it didn't exist; or insert the `original` value if it did.
    set!(dict, original, key)
end
```

So, I prefer using another verb for the API for this PR.

> Both "modify" and "update" refer to the dict rather than the value AFAICT

@nalimilan But doesn't `get` refer to a "slot" rather than the dict?  So, isn't it strange that the mutation API refers to the whole dict rather than a slot?  Also, other mutation verbs like `setindex!` and `setproperty!` seem to refer to a slot.

Reply:
> @clarkevans I think @nalimilan mentioned a good point that `set!` sounds like a counter part of `get`. However, I don't think it's a good parallel. If you consider `set!` as a counter part of `get(::Container, ::Key) ::Union{Nothing,Some{Value}}`, the signature of `set!` should be `set!(::Container, ::Union{Nothing,Some{Value}}, ::Key)`. This is how `getindex` and `setindex!` work.

Well currently we have `get(f::Function, collection, key)`, whose signature is very the same as `modify!(f, d, key)` in this PR.  If we added `get(collection, key)::Union{Nothing,Some}` it could make sense to add `set!(collection, value::Union{Nothing,Some}}, key)` too.



> @nalimilan But doesn't `get` refer to a "slot" rather than the dict? So, isn't it strange that the mutation API refers to the whole dict rather than a slot? Also, other mutation verbs like `setindex!` and `setproperty!` seem to refer to a slot.

I find it hard to tell TBH. It's hard to argue about these things, and if we wanted a fully consistent naming scheme maybe `get` should be called `getkey`...

Reply:
From a readability standpoint, if naively read `set!` I'd probably expect the key to exist after the operation. Which is also just like `get!`. I'd feel surprised if an operation named `set!` deleted something, to be honest!

> it could make sense to add

I suppose a simpler `modify!(d, key, v::Union{Nothing, Some{Value}})` could be nice. I note in other places in `Base` we do seem to be struggling whether dispatching on `::Function` is desirable and these arguments seemed to be getting widened to `::Any`. Partly because some callable things are `Function` or `Callable` (like python objects). (I was also sensing that there seemed to be some more general arguments that we should be able to know the semantics of a method by the function name and number of arguments only.)

> But doesn't get refer to a "slot" rather than the dict?

I agree - I have always read these verbs as refering to the "slot" on the container.

Reply:
How about [`crud!`](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) it does create, update, delete... and retrieve. There is also [`merge!`](https://en.wikipedia.org/wiki/Merge_(SQL)) aka `upsert!` from SQL.  I think `crud!` is a new kind of word that is fun to say and perhaps reflects this hybrid semantics?